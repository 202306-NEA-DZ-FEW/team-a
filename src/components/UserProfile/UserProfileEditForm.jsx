import { doc, updateDoc } from "firebase/firestore";
import { useFormik } from "formik";
import { useTranslation } from "next-i18next";
import { toast } from "react-toastify";
import * as Yup from "yup";

import { db } from "@/lib/firebase";
import getAllStates from "@/lib/getAllStates";

import Input from "../Input";
import SelectInput from "../SelectInput";

function UserProfileEditForm({ userData, onUpdate }) {
  const { t } = useTranslation();
  const states = getAllStates(t);

  const formik = useFormik({
    initialValues: {
      name: userData.name,
      email: userData.email,
      phone: userData.phone,
      location: userData.location,
    },
    validationSchema: Yup.object({
      name: Yup.string()
        .max(20, t("dashboard:userUpadteForm:nameLength"))
        .required(t("dashboard:userUpadteForm:nameRequired")),
      email: Yup.string()
        .email(t("dashboard:userUpadteForm:emailValid"))
        .required(t("dashboard:userUpadteForm:emailRequired")),
      phone: Yup.number().required(t("dashboard:userUpadteForm:phoneRequired")),
      location: Yup.string().required(
        t("dashboard:userUpadteForm:locationRequired")
      ),
    }),
    onSubmit: async (values) => {
      const userDocRef = doc(db, "users", userData.uid);
      try {
        await updateDoc(userDocRef, values);
        onUpdate({
          ...values,
          photoURL: userData.photoURL,
          uid: userData.uid,
          role: userData.role,
          date: userData.date,
        });
        toast.success("Profile updated successfully âœ…", {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
      } catch (error) {
        toast.error(error.code, {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
      } finally {
        document.getElementById("my_modal_1").close();
      }
    },
  });

  return (
    <dialog id='my_modal_1' className='modal bg-black bg-opacity-40'>
      <div className='modal-box bg-white transition-all duration-500 ease-in-out'>
        <form onSubmit={formik.handleSubmit} className='flex flex-col gap-4'>
          <h1 className='text-xl font-bold text-center'>
            Update your personal profile
          </h1>
          <Input
            name='name'
            type='text'
            placeholder={t("dashboard:userUpadteForm:namePlaceholder")}
            label={t("dashboard:userUpadteForm:nameLabel")}
            handleChange={formik.handleChange}
            value={formik.values.name}
            handleBlur={formik.handleBlur}
            error={formik.errors.name}
            touched={formik.touched.name}
          />
          <Input
            name='email'
            type='email'
            placeholder={t("dashboard:userUpadteForm:emailPlaceholder")}
            label={t("dashboard:userUpadteForm:emailLabel")}
            handleChange={formik.handleChange}
            value={formik.values.email}
            handleBlur={formik.handleBlur}
            error={formik.errors.email}
            touched={formik.touched.email}
          />
          <Input
            name='phone'
            type='text'
            placeholder={t("dashboard:userUpadteForm:phonePlaceholder")}
            label={t("dashboard:userUpadteForm:phoneLabel")}
            handleChange={formik.handleChange}
            value={formik.values.phone}
            handleBlur={formik.handleBlur}
            error={formik.errors.phone}
            touched={formik.touched.phone}
          />
          <SelectInput
            name='location'
            data={states}
            label={t("dashboard:userUpadteForm:locationLabel")}
            handleChange={formik.handleChange}
            value={formik.values.location}
            touched={formik.touched.location}
            error={formik.errors.location}
          />
          <div className='modal-action'>
            <button
              type='submit'
              className='btn rounded-full normal-case tracking-wider btn-active'
            >
              {t("dashboard:userUpadteForm:submit")}
            </button>
            <div method='dialog'>
              <button
                type='button'
                className='btn btn-outline rounded-full normal-case tracking-wider font-light'
                onClick={() => document.getElementById("my_modal_1").close()}
              >
                {t("dashboard:userUpadteForm:cancel")}
              </button>
            </div>
          </div>
        </form>
      </div>
    </dialog>
  );
}

export default UserProfileEditForm;
