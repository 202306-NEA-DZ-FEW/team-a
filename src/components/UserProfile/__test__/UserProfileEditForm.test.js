import renderer from "react-test-renderer";

import UserProfileEditForm from "../UserProfileEditForm";

jest.mock("@/lib/firebase", () => {
  return {
    app: {}, // Mock app object
    db: {}, // Mock db object
    storage: {}, // Mock storage object
    auth: {}, // Mock auth object
  };
});

it("renders correctly", () => {
  const mockUser = {
    name: "John",
    email: "jhm@gmail.com",
    phone: "123",
    location: "1- Adrad",
  };

  const mockOnUpdate = jest.fn();
  const tree = renderer
    .create(<UserProfileEditForm userData={mockUser} onUpdate={mockOnUpdate} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
