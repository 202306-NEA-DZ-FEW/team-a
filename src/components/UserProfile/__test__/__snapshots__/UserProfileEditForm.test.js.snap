// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<dialog
  className="modal"
  id="my_modal_1"
>
  <div
    className="modal-box"
  >
    <form
      className="flex flex-col gap-4"
      onSubmit={[Function]}
    >
      <h1
        className="text-3xl font-black text-center"
      >
        Update your personal profile
      </h1>
      <div>
        <div
          className="label ml-2"
        >
          <label
            className="label-text"
            htmlFor="name"
          >
            userUpadteForm.nameLabel
          </label>
          <label
            className="label-text-alt "
          />
        </div>
        <input
          className="input bg-white input-primary w-full rounded-full border-opacity-25"
          id="name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="userUpadteForm.namePlaceholder"
          type="text"
          value="John"
        />
      </div>
      <div>
        <div
          className="label ml-2"
        >
          <label
            className="label-text"
            htmlFor="email"
          >
            userUpadteForm.emailLabel
          </label>
          <label
            className="label-text-alt "
          />
        </div>
        <input
          className="input bg-white input-primary w-full rounded-full border-opacity-25"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="userUpadteForm.emailPlaceholder"
          type="email"
          value="jhm@gmail.com"
        />
      </div>
      <div>
        <div
          className="label ml-2"
        >
          <label
            className="label-text"
            htmlFor="phone"
          >
            userUpadteForm.phoneLabel
          </label>
          <label
            className="label-text-alt "
          />
        </div>
        <input
          className="input bg-white input-primary w-full rounded-full border-opacity-25"
          id="phone"
          name="phone"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="userUpadteForm.phonePlaceholder"
          type="text"
          value="123"
        />
      </div>
      <div>
        <div
          className="label"
        >
          <label
            className="label-text"
            htmlFor="location"
          >
            userUpadteForm.locationLabel
          </label>
          <label
            className="label-text-alt "
          />
        </div>
        <select
          className="select select-primary bg-white w-full border-opacity-25 rounded-3xl"
          defaultValue="1- Adrad"
          id="location"
          onChange={[Function]}
        >
          <option
            value="adrar"
          >
            adrar
          </option>
          <option
            value="chlef"
          >
            chlef
          </option>
          <option
            value="laghouat"
          >
            laghouat
          </option>
          <option
            value="oumElBouaghi"
          >
            oumElBouaghi
          </option>
          <option
            value="batna"
          >
            batna
          </option>
          <option
            value="bejaia"
          >
            bejaia
          </option>
          <option
            value="biskra"
          >
            biskra
          </option>
          <option
            value="bechar"
          >
            bechar
          </option>
          <option
            value="blida"
          >
            blida
          </option>
          <option
            value="bouira"
          >
            bouira
          </option>
          <option
            value="tamanrasset"
          >
            tamanrasset
          </option>
          <option
            value="tebessa"
          >
            tebessa
          </option>
          <option
            value="tlemcen"
          >
            tlemcen
          </option>
          <option
            value="tiaret"
          >
            tiaret
          </option>
          <option
            value="tiziOuzou"
          >
            tiziOuzou
          </option>
          <option
            value="algiers"
          >
            algiers
          </option>
          <option
            value="djelfa"
          >
            djelfa
          </option>
          <option
            value="jijel"
          >
            jijel
          </option>
          <option
            value="setif"
          >
            setif
          </option>
          <option
            value="saida"
          >
            saida
          </option>
          <option
            value="skikda"
          >
            skikda
          </option>
          <option
            value="sidiBelAbbes"
          >
            sidiBelAbbes
          </option>
          <option
            value="annaba"
          >
            annaba
          </option>
          <option
            value="guelma"
          >
            guelma
          </option>
          <option
            value="constantine"
          >
            constantine
          </option>
          <option
            value="medea"
          >
            medea
          </option>
          <option
            value="mostaganem"
          >
            mostaganem
          </option>
          <option
            value="mSila"
          >
            mSila
          </option>
          <option
            value="mascara"
          >
            mascara
          </option>
          <option
            value="ouargla"
          >
            ouargla
          </option>
          <option
            value="oran"
          >
            oran
          </option>
          <option
            value="elBayadh"
          >
            elBayadh
          </option>
          <option
            value="illizi"
          >
            illizi
          </option>
          <option
            value="bordjBouArreridj"
          >
            bordjBouArreridj
          </option>
          <option
            value="boumerdes"
          >
            boumerdes
          </option>
          <option
            value="elTarf"
          >
            elTarf
          </option>
          <option
            value="tindouf"
          >
            tindouf
          </option>
          <option
            value="tissemsilt"
          >
            tissemsilt
          </option>
          <option
            value="elOued"
          >
            elOued
          </option>
          <option
            value="khenchela"
          >
            khenchela
          </option>
          <option
            value="soukAhras"
          >
            soukAhras
          </option>
          <option
            value="tipaza"
          >
            tipaza
          </option>
          <option
            value="mila"
          >
            mila
          </option>
          <option
            value="ainDefla"
          >
            ainDefla
          </option>
          <option
            value="naama"
          >
            naama
          </option>
          <option
            value="ainTemouchent"
          >
            ainTemouchent
          </option>
          <option
            value="ghardaia"
          >
            ghardaia
          </option>
          <option
            value="relizane"
          >
            relizane
          </option>
          <option
            value="timimoun"
          >
            timimoun
          </option>
          <option
            value="bordjBadjiMokhtar"
          >
            bordjBadjiMokhtar
          </option>
          <option
            value="ouledDjellal"
          >
            ouledDjellal
          </option>
          <option
            value="beniAbbes"
          >
            beniAbbes
          </option>
          <option
            value="ainSalah"
          >
            ainSalah
          </option>
          <option
            value="ainGuezzam"
          >
            ainGuezzam
          </option>
          <option
            value="touggourt"
          >
            touggourt
          </option>
          <option
            value="djanet"
          >
            djanet
          </option>
          <option
            value="elMGhair"
          >
            elMGhair
          </option>
          <option
            value="elMenia"
          >
            elMenia
          </option>
        </select>
      </div>
      <div
        className="modal-action"
      >
        <button
          className="btn btn-primary"
          type="submit"
        >
          userUpadteForm.submit
        </button>
        <div
          method="dialog"
        >
          <button
            className="btn"
            onClick={[Function]}
            type="button"
          >
            userUpadteForm.cancel
          </button>
        </div>
      </div>
    </form>
  </div>
</dialog>
`;
